### ISSUE TITLE
Can it be used in conjunction with Streamlit?

### ISSUE THREAD

**Issue by hanxuanliang** [louisnw01/lightweight-charts-python#1]

**Comment by louisnw01** [comment]
Hi Han,

**Comment by louisnw01** [comment]
Hi Han,

To my knowledge Streamlit doesn't allow Javascript to be dynamically updated, so (unless i am wrong) the live data features of this library will not be compatible.

However, static data can be displayed in streamlit, as shown in the following example which uses the 'LWC' class of lightweight-charts-python: 

```python
import pandas as pd
from streamlit.components.v1 import html

from lightweight_charts import LWC


class StreamlitChart(LWC):
    def __init__(self, volume_enabled=True, width=None, height=None):
        super().__init__(volume_enabled)
        self.loaded = True

        self.width = width
        self.height = height

        self._html = self._html.replace('</script>\n</body>\n</html>', '')

    def run_script(self, script): self._html += '\n'+script

    def load(self): html(f'{self._html}</script></body></html>', width=self.width, height=self.height)


def main():
    chart = StreamlitChart(width=900, height=600)

    df = pd.read_csv('ohlcv.csv')
    chart.set(df)

    chart.load()


if __name__ == "__main__":
    main()
```

In this example, all of the calls to change the chart need to happen before `chart.load()` is called, which means that anything static, ie non-changing data, will work fine. Anything using `chart.update()`, `chart.update_from_tick()`, or calls after the `chart.load()`  will not work.

**Comment by hanxuanliang** [comment]
```py
import pandas as pd
from model.stock import get_trade_data
import MyTT as mt

from lightweight_charts import Chart, LWC
from streamlit.components.v1 import html


class StreamlitChart(LWC):
    def __init__(self, volume_enabled=True, width=None, height=None):
        super().__init__(volume_enabled)
        self.loaded = True

        self.width = width
        self.height = height

        self._html = self._html.replace('</script>\n</body>\n</html>', '')

    def run_script(self, script): self._html += '\n'+script

    def load(self): html(
        f'{self._html}</script></body></html>', width=self.width, height=self.height)


def rende_data():
    chart = StreamlitChart(width=900, height=600)

    # get your trade data
    trade_data = get_trade_data('688111', level='1d')
    trade_data.rename(columns={'trade_time': 'date'}, inplace=True)
    trade_data.index = pd.to_datetime(trade_data['date'])
    chart.set(trade_data)

    trade_data['ma_233'] = mt.SMA(trade_data["close"], 233)
    trade_data['ma_55'] = mt.SMA(trade_data["close"], 55)

    # => the question code?
    # line1 = chart.create_line()
    # sma_233_data = [{'time': item['date'], 'value': item['ma_233']}
    #                 for _, item in trade_data.iterrows()]
    # line1.set(pd.DataFrame(sma_233_data))

    line2 = chart.create_line()
    sma_55_data = [{'time': item['date'], 'value': item['ma_55']}
                    for _, item in trade_data.iterrows()]
    line2.set(pd.DataFrame(sma_55_data))

    chart.load()


if __name__ == "__main__":
    rende_data()
```

Thank you to the developers. Yes, following your prompts, I can now accurately depict the data.
However, a new problem has arisen: How do I render two lines?
It seems that when I release the commented-out code, Streamlit is unable to display both curves.

**Comment by louisnw01** [comment]
It appears this issue is because in JavaScript a variable is being declared twice with the same name. A fix has been found, and it will (hopefully) be published tommorow.

Apologies!
Louis

**Comment by hanxuanliang** [comment]
Great! 🚀

**Comment by louisnw01** [comment]
> Great! 🚀

This issue has now been fixed and published in the latest version (1.0.4).
Let me know if you're still running into any issues!

**Comment by hanxuanliang** [comment]
I'm here to ask a question again. Hhhh. 😊

In the latest version 1.0.7, the code for `StreamlitChart(LWC)` will render as a black screen (just a black background without rendering data). Then I rolled back to version 1.0.6, which runs normally.

v1.0.6
![image](https://github.com/louisnw01/lightweight-charts-python/assets/29874640/4f4f899c-b6f8-445b-b6c9-bc21ef77c58f)

v1.0.7
![image](https://github.com/louisnw01/lightweight-charts-python/assets/29874640/e513b511-d87d-4255-915d-4dd3cb47dbe6)

**Comment by hanxuanliang** [comment]
It seems there is still a problem. The data has not been rendered, and I have discovered a new issue: the rendering time is too long.

<img width="1456" alt="image" src="https://github.com/louisnw01/lightweight-charts-python/assets/29874640/be85c5f1-0875-4433-9268-e0ae2306ca24">

**Comment by louisnw01** [comment]
Does the data render at all, or just take too long to load?

How long is the DataFrame you are using? At the moment, there is an issue with speed when it comes to processing large datasets. The next update will contain an optimisation to mitigate this however.

**Comment by louisnw01** [comment]
I should add; I am unable to reproduce this error with the example I provided.

**Comment by hanxuanliang** [comment]
> Does the data render at all, or just take too long to load?
> 
> How long is the DataFrame you are using? At the moment, there is an issue with speed when it comes to processing large datasets. The next update will contain an optimisation to mitigate this however.

The size of the entire dataset is: daily data from 1996-01-01 to 2023-05-22. However, currently this issue seems to only occur in larger datasets and may render slowly. This problem did not appear in version 1.0.6.
However, this issue is not very urgent as version 1.0.6 works very well and I have been using it for some time now. You can prioritize more urgent issues.

**Comment by louisnw01** [comment]
Hey again,

The latest version (1.0.9) now contains the StreamlitChart object. Same functionality, you just don't need to write out the class verbose anymore:

```python
from lightweight_charts.widgets import StreamlitChart
```
